### ISSUE TITLE
investiage tensorflow

### ISSUE THREAD

**Issue by yorkerlin** [shogun-toolbox/shogun#3655]
@karlnapf @vigsterkr  @lisitsyn 
It seems that tensorflow now supports to build a computational graph in c++.
https://www.tensorflow.org/api_guides/cc/guide

Do we create a tensorflow backend for the linear algebra lib or GP?

**Comment by yorkerlin** [comment]
Alternatively, we could also make use of this project, https://github.com/dmlc/nnvm.

**Comment by lisitsyn** [comment]
The most important is Tensoflow Serving (https://tensorflow.github.io/serving/) that would enable us to use trained models within Shogun.

**Comment by vigsterkr** [comment]
XLA https://haosdent.gitbooks.io/tensorflow-document/content/resources/xla_prerelease.html

**Comment by yorkerlin** [comment]
Nice! I will wait for XLA.

**Comment by karlnapf** [comment]
Step one: Prototype a Shogun kernel that uses tensorflow to do computations. There is 1000 things to figure out and this covers quite a few of them.
Step two: Compute the kernel gradients with tensorflow
Step three: think about how we can apply this shogun - wide.

Questions:
 * Is `linalg` the right place in Shogun to integrate this into?
 * Do we want yet another API for graph computations?
 * XLS is very early alpha, shall we wait a bit longer?
 * Autodiff is still very  useful, which doesn't need XLA, but just an autodiff API

**Comment by vigsterkr** [comment]
@karlnapf started a gluing lib with tf serving, i guess later using the same idea one can do tighter integration with tf.

**Comment by yorkerlin** [comment]
@karlnapf  Yes. One approach is to read code from https://github.com/GPflow/GPflow and get some ideas from the project.

**Comment by yorkerlin** [comment]
BTW, there is a benchmark for GPs in python. https://github.com/GPflow/GPflowBenchmarks

**Comment by karlnapf** [comment]
Should be done using the benchmarking framework by mlpack: https://github.com/zoq/benchmarks
